In [3]:
library(tidyverse)
library(dplyr)
library(tidymodels)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
── Attaching packages ────────────────────────────────────── tidymodels 1.1.1 ──

✔ broom        1.0.6     ✔ rsample      1.2.1
✔ dials        1.3.0     ✔ tune         1.1.2
✔ infer        1.0.7     ✔ workflows    1.1.4
✔ modeldata    1.4.0     ✔ workflowsets 1.0.1
✔ parsnip      1.2.1     ✔ yardstick    1.3.1
✔ recipes      1.1.0     

── Conflicts ───────────────────────────────────────── tidymodels_conflicts() ──
✖ scales::discard() masks purrr::discard()
✖ dplyr::filt

In [4]:
#Codes for The Players Data Set:

players_data <- read_csv("data/players.csv")

#Finding out How many Observations
number_observations_players <- nrow(players_data)
number_observations_players

#Table for variables meaning and type
variables_players <- tibble::tibble(
  Variable = c("experience", "subscribe", "hashedEmail", 
               "played_hours", "name", "gender", "Age"),
  
  Meaning = c("Player’s experience/skill level",
              "Indicates if the player is a subscriber to the game's (MineCraft) newsletter",
              "Anonymized unique identifier for each player",
              "Total number of hours the player has played",
              "Player’s first name",
              "Player’s gender identity",
              "Player’s age in years"),
  
  Type = c("Categorical", "Categorical", "Categorical",
           "Quantitative", "Categorical", "Categorical", "Quantitative"))

#Percentage of players' skill levels
experience_players <- players_data |>
count(experience)|> mutate(Percent_of_Skill_Levels_Among_Players = round(n/196*100, 2)) 


#Percentage of players' are subscribed or not
subscribe_players <- players_data |> count(subscribe) |> mutate(Percent_of_Subscription_Status = round(n/196*100, 2)) 

#Count and Percentage of 
gender_players <- players_data |> count(gender) |> mutate(Percent_of_Genders = round(n/196 *100, 2))

#Summary Statistics of Variables
summary(players_data)

Rows: 196 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (4): experience, hashedEmail, name, gender
dbl (2): played_hours, Age
lgl (1): subscribe

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


[1] 196

  experience        subscribe       hashedEmail         played_hours    
 Length:196         Mode :logical   Length:196         Min.   :  0.000  
 Class :character   FALSE:52        Class :character   1st Qu.:  0.000  
 Mode  :character   TRUE :144       Mode  :character   Median :  0.100  
                                                       Mean   :  5.846  
                                                       3rd Qu.:  0.600  
                                                       Max.   :223.100  
                                                                        
     name              gender               Age       
 Length:196         Length:196         Min.   : 9.00  
 Class :character   Class :character   1st Qu.:17.00  
 Mode  :character   Mode  :character   Median :19.00  
                                       Mean   :21.14  
                                       3rd Qu.:22.75  
                                       Max.   :58.00  
                               

In [5]:
#Codes for the Sessions data set
sessions_data <- read_csv("data/sessions.csv")
head(sessions_data)

#Finding out How many Observations
number_observations_sessions <- nrow(sessions_data)
number_observations_sessions

#Number of Variables and their Names
colnames(sessions_data)

#Table for Variables meaning and Type
variables_sessions <- tibble::tibble(
  Variable = c("hashedEmail", "start_time", "end_time", 
               "original_start_time", "original_end_time"),
  
  Meaning = c("Anonymized unique identifier for each player",
              "The actual recorded start time of a session", 
              "The actual recorded end time of a session", 
              "Original start timestamp (ms)", 
              "Original end timestamp (ms)"),
  
  Type = c("Categorical", "Quantitative", "Quantitative",
           "Quantitative", "Quantitative"))

summary(sessions_data)

Rows: 1535 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): hashedEmail, start_time, end_time
dbl (2): original_start_time, original_end_time

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


hashedEmail                                                     
1 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
2 36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
3 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc
4 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
5 36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686
6 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf
  start_time       end_time         original_start_time original_end_time
1 30/06/2024 18:12 30/06/2024 18:24 1.71977e+12         1.71977e+12      
2 17/06/2024 23:33 17/06/2024 23:46 1.71867e+12         1.71867e+12      
3 25/07/2024 17:34 25/07/2024 17:57 1.72193e+12         1.72193e+12      
4 25/07/2024 03:22 25/07/2024 03:58 1.72188e+12         1.72188e+12      
5 25/05/2024 16:01 25/05/2024 16:12 1.71665e+12         1.71665e+12      
6 23/06/2024 15:08 23/06/2024 17:10 1.71916e+12         1.71916e+12

[1] 1535

[1] "hashedEmail"         "start_time"          "end_time"           
[4] "original_start_time" "original_end_time"

 hashedEmail         start_time          end_time         original_start_time
 Length:1535        Length:1535        Length:1535        Min.   :1.712e+12  
 Class :character   Class :character   Class :character   1st Qu.:1.716e+12  
 Mode  :character   Mode  :character   Mode  :character   Median :1.719e+12  
                                                          Mean   :1.719e+12  
                                                          3rd Qu.:1.722e+12  
                                                          Max.   :1.727e+12  
                                                                             
 original_end_time  
 Min.   :1.712e+12  
 1st Qu.:1.716e+12  
 Median :1.719e+12  
 Mean   :1.719e+12  
 3rd Qu.:1.722e+12  
 Max.   :1.727e+12  
 NA's   :2          

<h2> <u> (1) Data Description:</u> </h2>

<u>**Players Data Set Description:**</u>

A comma-separated-values data set including information about game players. It has 196 observations and 7 variables.

In [5]:
variables_players

Variable    
1 experience  
2 subscribe   
3 hashedEmail 
4 played_hours
5 name        
6 gender      
7 Age         
  Meaning                                                                     
1 Player’s experience/skill level                                             
2 Indicates if the player is a subscriber to the game's (MineCraft) newsletter
3 Anonymized unique identifier for each player                                
4 Total number of hours the player has played                                 
5 Player’s first name                                                         
6 Player’s gender identity                                                    
7 Player’s age in years                                                       
  Type        
1 Categorical 
2 Categorical 
3 Categorical 
4 Quantitative
5 Categorical 
6 Categorical 
7 Quantitative

The **experience** variable includes five skill levels. The percentage and count (n) of each skill level is described below.



In [7]:
experience_players

experience n  Percent_of_Skill_Levels_Among_Players
1 Amateur    63 32.14                                
2 Beginner   35 17.86                                
3 Pro        14  7.14                                
4 Regular    36 18.37                                
5 Veteran    48 24.49

The **subscribe** variable's summary statistics is described below. The table indicates "TRUE" meaning they have subscribed and "FALSE" meaning they have not and "n" column as the count.

In [8]:
subscribe_players

subscribe n   Percent_of_Subscription_Status
1 FALSE      52 26.53                         
2  TRUE     144 73.47

For **played hours**, the mean hours played among the 196 players is 5.85 hours. A possible issue with this is that hours played for the players could have outliers, being too different from the common trend in the data, which distorts the mean and could pose as an issue for linear regression models who are sensitive to outliers.

There are seven categories listed for **gender**. The count (n) and percentages of genders are listed below. The mean age of the players is 21. If used as a predictor, the "Other" and "Prefer not to say" categories can pose an issue for K-NN classification models. If the optimal K chosen includes the majority of the neighbors being one of these, an idea of what gender classifies the label may not be formed.

In [9]:
gender_players

gender            n   Percent_of_Genders
1 Agender             2  1.02             
2 Female             37 18.88             
3 Male              124 63.27             
4 Non-binary         15  7.65             
5 Other               1  0.51             
6 Prefer not to say  11  5.61             
7 Two-Spirited        6  3.06

The players **ages** range from 9 years old to 58 years old with a mean of 21 years old. This wide range of ages could be an issue as the extreme values could be outliers and distort the mean and pose an issue for linear regression models as well.

Furthermore, a potential issue could arise if the data was collected with players personally inputting their information. The players could input wrong information, such as distorting their age due to age restrictions for the game. 

<u>**Sessions Data Set Description**</u>

A comma-separated-values data set including a list of data about individual play sessions of each game player. It has 1535 observations and 5 variables.

In [6]:
variables_sessions

Variable            Meaning                                      Type        
1 hashedEmail         Anonymized unique identifier for each player Categorical 
2 start_time          The actual recorded start time of a session  Quantitative
3 end_time            The actual recorded end time of a session    Quantitative
4 original_start_time Original start timestamp (ms)                Quantitative
5 original_end_time   Original end timestamp (ms)                  Quantitative

* Summary Statistics:
  
  After computing the duration of each session using the data, the mean is 50.86 minutes. 

In [7]:
#Computation of Mean Session Duration: 
# (code used is not from class, referenced from the "AI Overview" section when searched on Google)

#Using the clock time to calculate session duration in minutes
session_durations <- sessions_data |> mutate(
    start_time = dmy_hm(start_time),  # day-month-year hour:minute
    end_time = dmy_hm(end_time),
    session_duration_mins = as.numeric(difftime(end_time, start_time, units = "mins"))
  )

#Calculating the mean
summary(session_durations)

 hashedEmail          start_time                    
 Length:1535        Min.   :2024-04-06 09:27:00.00  
 Class :character   1st Qu.:2024-05-20 20:28:00.00  
 Mode  :character   Median :2024-06-24 02:51:00.00  
                    Mean   :2024-06-24 03:54:03.51  
                    3rd Qu.:2024-07-25 06:56:00.00  
                    Max.   :2024-09-26 06:09:00.00  
                                                    
    end_time                      original_start_time original_end_time  
 Min.   :2024-04-06 09:31:00.00   Min.   :1.712e+12   Min.   :1.712e+12  
 1st Qu.:2024-05-20 21:37:00.00   1st Qu.:1.716e+12   1st Qu.:1.716e+12  
 Median :2024-06-23 22:04:00.00   Median :1.719e+12   Median :1.719e+12  
 Mean   :2024-06-24 02:26:18.03   Mean   :1.719e+12   Mean   :1.719e+12  
 3rd Qu.:2024-07-25 07:00:00.00   3rd Qu.:1.722e+12   3rd Qu.:1.722e+12  
 Max.   :2024-09-26 07:39:00.00   Max.   :1.727e+12   Max.   :1.727e+12  
 NA's   :2                                            NA's

* Issues in the Data:
  
  An issue with the data is that it is not tidy because the columns start_time and end_time have two values, date and time, in a single cell. 

In [8]:
sessions_data_tidy <- sessions_data |> 
separate(col = start_time, 
         into = c("start_Date", "start_time"),
         sep = " ") |>
separate(col = end_time, 
         into = c("end_date", "end_time"),
         sep = " ")
head(sessions_data_tidy)  

hashedEmail                                                      start_Date
1 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf 30/06/2024
2 36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686 17/06/2024
3 f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3398304c7ae42581fdc 25/07/2024
4 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf 25/07/2024
5 36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f575d4acc9cf487c4686 25/05/2024
6 bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431d8aa0c4bf95ccee6bf 23/06/2024
  start_time end_date   end_time original_start_time original_end_time
1 18:12      30/06/2024 18:24    1.71977e+12         1.71977e+12      
2 23:33      17/06/2024 23:46    1.71867e+12         1.71867e+12      
3 17:34      25/07/2024 17:57    1.72193e+12         1.72193e+12      
4 03:22      25/07/2024 03:58    1.72188e+12         1.72188e+12      
5 16:01      25/05/2024 16:12    1.71665e+12         1.71665e+12      
6 15:08      23/06/2024 17:10    1.71916e+12         1.71916e+12

* Exploration of Variables

  to be finished...

<h2> <u> (2) Questions </u> </h2>